multidimensional_sampling/pgen_sample.py

```diff
@@ -0,0 +1,449 @@
+        constants = {}
+        for key, value in sampling_dict["constants"].items():
+            constants[key] = value
+    else:
+        constants = {}
+    if "parameters" in sampling_dict.keys():
+        parameters = sampling_dict["parameters"].copy()
+        samples = _recursive_cross_product_sample(
+                                            parameters,
+                                            samples=[constants])
+    else:
+        samples = [constants]
+    LOGGER.info("samples:\n%s", str(samples))
+    return samples
+
+
+def _validate_column_list_dictionary(sampling_dict):
+    _validate_constants_parameters(sampling_dict)
+    _validate_parameters_string(sampling_dict)
+
+
+def column_list_sample(sampling_dict):
+    """
+    Return set of samples based on specification in sampling_dict.
+
+    Prototype dictionary:
+
+    sample_type: column_list
+    constants:
+        X3: 20
+    parameters: |
+        X1       X2
+        5        10
+        3        7
+        12       16
+    """
+    _validate_column_list_dictionary(sampling_dict)
+    samples = []
+    if "constants" in sampling_dict.keys():
+        constants = {}
+        for key, value in sampling_dict["constants"].items():
+            constants[key] = value
+    else:
+        constants = {}
+    if "parameters" in sampling_dict.keys():
+        rows = sampling_dict["parameters"].split('\n')
+        headers = rows.pop(0).split()
+        for row in rows:
+            data = row.split()
+            if len(data) > 0:
+                _log_assert(
+                    len(data) == len(headers),
+                    "Data >>" + str(data) + "<< does not match\n" +
+                    "headers >>" + str(headers) + "<<.")
+                sample = constants.copy()
+                for header, datum in zip(headers, data):
+                    sample[header] = datum
+                samples.append(sample)
+    else:
+        samples.append(constants)
+    return samples
+
+
+def _validate_best_candidate_dictionary(sampling_dict):
+    _validate_constants_parameters(sampling_dict)
+    _validate_parameters_dict(sampling_dict)
+    _log_assert(
+        type(sampling_dict["num_samples"]) == int,
+        "'num_samples' must exist and be an integer")
+    _log_assert(
+        isinstance(sampling_dict["parameters"], dict),
+        "'parameters' must exist and be a dictionary")
+    if 'previous_samples' in sampling_dict.keys():
+        pass
+        # TTD: validate that file exists and that it
+        # contains same parameters as `parameters`
+    for key, value in sampling_dict["parameters"].items():
+        _log_assert(type(key) == str, "parameter labels must be strings")
+        _log_assert(
+            str(value["min"]).isnumeric(),
+            "parameter must have a numeric minimum")
+        _log_assert(
+            str(value["max"]).isnumeric(),
+            "parameter must have a numeric maximum")
+
+def down(samples, sampling_dict):

```

Comment:
It may be better to explicitly name this function `down_select` or something similar.